### PR TITLE
feat: support preview colorscheme when move in colors wildmenu

### PR DESCRIPTION
**Pull request by glepnir**
not complete yet. do we need this?

colorscheme

### CODE REVIEW

**Comment by clason** [comment]
Neat! But isn't that something that could be contributed to Vim first?

**Comment by glepnir** [comment]
there have some code and workflow i can't understand, need look more code of there. I will try to complete it in neovim first then will try to copy to vim. thanks for advice

File: src/nvim/cmdexpand.c [hunk 1]
```diff
@@ -853,11 +853,18 @@ char *ExpandOne(expand_T *xp, char *str, char *orig, int options, int mode)
   if (mode == WILD_NEXT || mode == WILD_PREV
       || mode == WILD_PAGEUP || mode == WILD_PAGEDOWN
       || mode == WILD_PUM_WANT) {
-    return get_next_or_prev_match(mode, xp, &findex, orig_save);
+    char *item = get_next_or_prev_match(mode, xp, &findex, orig_save);
+    if (xp->xp_context == EXPAND_COLORS && strcmp(item,"default") != 0) {
```
Comment (zeertzjq) [inline src/nvim/cmdexpand.c:857]: Why special-case "default"?

Comment (glepnir) [reply src/nvim/cmdexpand.c:857]: a test code nvm

Comment (glepnir) [reply src/nvim/cmdexpand.c:857]: a test code nvm

**Comment by lewis6991** [comment]
I don't like the idea of baking this into `cmdexpand.c`. Could we not achieve this higher up by invoking a completion callback for `:color` with the use of `:noautocmd`?

**Comment by glepnir** [comment]
also can do it in `ex_getln.c`  `command_line_wildchar_complete` function.  not sure where is best. `trigger_event` is dirty .. but curently i don't find a good way to do avoid trigger event . as a plugin need implement a wildchar move callback ?  not all command need this callback... because some editor support this feature as builtin . so try to implement it

**Comment by lewis6991** [comment]
Complete implementation:

```lua
local orig_color
vim.api.nvim_create_autocmd({'CmdlineEnter', 'CmdLineChanged', 'CmdlineLeave'}, {
  callback = function(data)
    if data.event == 'CmdlineEnter' then
      orig_color = vim.g.colors_name
    elseif data.event == 'CmdlineLeave' then
      vim.cmd.color(orig_color)
    else
      local parts = vim.split(vim.fn.getcmdline(), '%s+')
      if vim.startswith(parts[1], 'color') then
        vim.o.eventignore = 'all'
        pcall(vim.cmd.color, parts[2])
        vim.o.eventignore = ''
      end
    end
  end
})
```

**Comment by glepnir** [comment]
I have this similar code  in my config. not sure which is best . this way do more hack.  you want make it as a plugin in core ?

**Comment by glepnir** [comment]
I have this similar code  in my config. not sure which is best . this way do more hack.  you want make it as a plugin in core ?

**Comment by lewis6991** [comment]
Why is it a hack? It is completely contained, doesn't involve any C and doesn't clutter up general C code with fragmented special case handling for a specific command.

**Comment by glepnir** [comment]
I just want to be able to preview the colorscheme like other editors when executing the colorscheme. Write callback in cmdline auto event. I don't find it elegant. and do it in core more simply no eventignore no command check no if else if pcall

**Comment by lewis6991** [comment]
It's not more elegant at all since you need to pollute generic code with command specific behaviours.

> no eventignore no command check no if else if pcall

None of this is bad.

**Comment by zeertzjq** [comment]
I don't the current implementation in this PR is correct. It will interfere with `getcompletion()`, and it also `-complete=color` user command completion, which may or may not be a good thing.

**Comment by glepnir** [comment]
IDK just first look at code of there.

**Comment by zeertzjq** [comment]
The current implementation in this PR makes the preview also apply to user commands with `-complete=color`, which may be unwanted.

**Comment by glepnir** [comment]
thanks for point it out.  just want to know the opinion is what . use lua as a plugin level implement it in core or c code in core. if c i will continue do this or implement a high level callback or other idea.  if lua close then.

**Comment by zeertzjq** [comment]
The previews in this PR don't fully work, because `update_screen()` is not called after loading the colorscheme. Calling `update_screen()` at an inappropriate place in cmdline mode can also cause problems.

**Comment by glepnir** [comment]
so it's draft I looking the source code of there.

**Comment by justinmk** [comment]
This should only activate for "interactive" use, does it check for that? Same for the Lua autocmd-based approach.

I doubt this will be accepted in vim, except under an option.

> if c i will continue do this or implement a high level callback or other idea. if lua close then.

I think we should avoid doing it in the core C path, that is too special-case. I agree that `vim.split(vim.fn.getcmdline()` is a hack but we could improve it later. To avoid this "hack" we need `v:event` to tell us what the parsed command is. Or I guess we could use `nvim_parse_cmd` ?

**Pull request closed without merging**